### DISCUSSION TITLE
Problem about using Moltres to solve an EigenValue benchmark

### DISCUSSION THREAD

**Discussion by js-jixu** [arfc/moltres#201] [Q&A Reactor multiphysics] [answered]
Hi, everyone.

I'm using Moltres to do a IAEA2D benchmark.  My understanding of this benchmark is to solve a two-group steady-state neutron diffusion equation for four materials and the group constants for the materials are given. First, given a hypothetical eigenvalue, the neutron flux is obtained by solving the diffusion equations. Then use the neutron flux calculated in the previous step to correct the new eigenvalues. This iteration goes on to get a final eigenvalue.

With @smpark7 's [advice](https://github.com/arfc/moltres/discussions/200), I  wrote a new input file today referring to the input file: **moltres/problems/2021-cnrs-benchmark/phase-0/nts.i**. The distribution of materials and corresponding group instants are here.

![materials](https://user-images.githubusercontent.com/99813361/166957969-982a4564-a0c8-40dd-9143-3ceb408d5f86.png)

The other group constants are all set zero, including reciprocal of speed of neutrons(INVV), delayed fission spectrum 
neutrons(CHID), average decay constant of delayed precursors in precursor group i(LAMBDA), energy produced per fission(KAPPA), fission cross section(FISS) and average neutrons produced per fission(NUBAR). Beacuse IAEA2D benchmark has no delayed neutrons and precursors, so I set number of precursors is 0. Then Moltres doesn't use promot fission spectrum(CHIP), but use total fission probability(CHIT). So I set them to the same value. The input file and group constant files are attatched in the archive. I set global temperature equals to 150 to get interpolated group constants. Other settings are as same as the referenced input file.

But the result is incorrect and strange, the hypothetical eigenvalue is 1.5, but the final eigenvalue is 276.7433. Do I set any wrong parameters in the input file?

![result](https://user-images.githubusercontent.com/99813361/166962244-39034df7-c93d-4191-97d4-8e132a181829.png)

Besides that input file, I found an input file: **moltres/problems/022317_test_critical_neutronics_only_reactor/k-eigen-problem.i**. And I plan to write a new input file by referencing this input file. Do you think this approach is feasible?

[Moltres_Discussion.zip](https://github.com/arfc/moltres/files/8633019/Moltres_Discussion.zip)

**Comment by smpark7** [comment, marked as answer]
Hey Xu. Your zip link is broken but I think you're on the right track with your current input file. Some of the issues I can think of are:
- Make sure your group constants are provided at the same temperature (150) that you set your global temperature. Moltres handles interpolation using spline or cubic spline between two provided values. It does not perform any complicated Doppler broadening calculations.
- `ElmIntegTotFissPostprocessor` requires `FISS` to calculate the total number of fissions, while `ElmIntegTotHeatPostprocessor` requires `KAPPA` to calculate the total power output. If you're using either of these postprocessors for the normalization factor in the `Executioner` section, you have to provide these group constant values. The rest of the group constants you mentioned can be set to zero.

You can choose not to provide any normalization parameters. The `InversePowerMethod` automatically normalizes to producing _k_ neutrons. Or you can use `ElmIntegTotNtsPostprocessor` to normalize by any number of fission neutrons you wish to produce.

**Reply by js-jixu** [reply]
Sorry, Sun Myung. I uploaded it again.

**Reply by js-jixu** [reply]
I set "temperature = 0" in the GlobalParams block. The literature attached in the folder seems that  has no information about FISS or KAPPA. Should I choose another simpler input file to modify. Just like the input file I mentioned at the end of the discussion.

**Reply by smpark7** [reply]
I would argue that `nts.i` is simpler than the other input file. There is nothing wrong with your input file. Your `REMXS` values here are incorrect. The removal cross section is the sum of the absorption cross section and the out-scatter cross sections. Your `NSF` value for mat4 should be zero too.

**Reply by js-jixu** [reply]
Hi Sun Myung.

I rebuilt the group constants folder `group_constants_sss2` and got new results. When I use the group constants in the literature and run the input file `benchmark_literature_group_constants.i` , the final eigenvalue is `3.1796136269` while the reference solution of this benchmark is `1.029585` . One of assistant professors in my research group told me that the removal cross section is not needed when solving the diffusion equation, only the absorption cross section is needed. Therefore, the group constants given by the benchmark questions can already meet the calculation requirements. 

![3 1796136](https://user-images.githubusercontent.com/99813361/167130536-1389fe0c-66b7-4ac9-9149-06dcc2cb3f63.png)

I have a friend from another university, who compiled a simple MOOSE-based diffusion equation calculation code(he calls it `Rabbit` ) with reference to Moltres. He used the same group constants and got same results, his input file `IAEA2D_WangSS_group_constants_modified.i` is attached. He said that his teacher once wrote a neutron diffusion equation calculation code based on MOOSE, and an undergraduate student used this code to do the IAEA2D benchmark problem, giving a different set of group constants in the undergraduate input file. This set of group constants is quite different from the group constants in the literature in the removal cross section and cross-section of scattering from group 1 to group 2. 

![group_constants_from_other_source](https://user-images.githubusercontent.com/99813361/167130605-81e5b589-610a-4669-9df3-23725bc03546.png)

My friend's another input file `IEAE2D_WangSS.i` used this set of group constants and got very accurate results——`1.029605` . The difference between this results and the reference solution is only 2 pcm. The output file `IAEA2D_WangSS_out.e` is attached. 

![right_output_group_1](https://user-images.githubusercontent.com/99813361/167130970-84a13cc3-c2fd-482c-92d3-75dbc7c17409.png)

Then I built a new group constants folder `group_constants_sss2_othersource` and run the input file `benchmark_other_source_group_constants` . But the final eigenvalue `0.0145448` is too small. The output file `benchmark_other_source_group_constants_outputs.e` is attached. 

![0 0145448](https://user-images.githubusercontent.com/99813361/167130688-80de1efd-411b-42ed-a989-5a1dba46adc5.png)

Since the results of the calculations are the same when the Moltres input file and the Rabbit input file use group constants from the literature, it means that the two input files are equivalent in a certain sense. However, when using the group constants of the undergraduate input file, the input file of Rabbit can calculate a result close to the reference solution, while the input file of Moltres obtains a result that is much different. 

These files are included in the zip: the folder of group constants from literature `group_constants_sss2` , the folder of group constants from other source(the undergraduate's input file) `group_constants_sss2_othersource` , the input file based on Moltres and group constants from literature `benchmark_literature_group_constants.i` , the input file based on Moltres and group constants from other source `benchmark_other_source_group_constants.i` and its corresponding output file `benchmark_other_source_group_constants_outputs.e` , IAEA2D geometry file `IAEA2D.geo` and mesh file `IAEA2D.msh` , the input file based on Rabbit and group constants from literature `IAEA2D_WangSS_group_constants_modified.i` , the input file based on Rabbit and group constants from other source `IAEA2D_WangSS.i` and its corresponding output file `IAEA2D_WangSS_out.e` . The IAEA2D_WangSS_out.e shows a solution very close to reference solution.

With best wishes, 

Xu.

[2022.05.06.zip](https://github.com/arfc/moltres/files/8639992/2022.05.06.zip)

**Reply by smpark7** [reply]
Hi Xu.

We designed Moltres to use the removal cross section instead of the absorption cross section based on the neutron diffusion equation derived in our [paper](https://www.sciencedirect.com/science/article/pii/S0306454917304760). This is simply a design choice. Others have written their neutron diffusion codes to use the absorption cross section. Both approaches are valid because the governing equations are derived from the standard multigroup neutron diffusion equation. We felt that using removal cross section is more convenient because Serpent provides them directly, or they can be easily pre-calculated otherwise. The MOOSE framework is a general and flexible framework which developers can use to develop their own code. Therefore, two MOOSE-based neutron diffusion codes written by different people can be very different in implementation.

There are two errors in your files:
1. Moltres requires the removal cross sections, which is the sum of the absorption cross section and the out-scatter cross section
2. Your scattering cross section is the wrong order of magnitude. They should be 0.02 or 0.04, instead of 0.2 and 0.4

When I corrected these errors, I got a k_eff of 1.034074 from the mesh file you provided. Here are the corrected files and the console output (`benchmark_modified_from_nts.out`):

[corrected_files.zip](https://github.com/arfc/moltres/files/8642635/corrected_files.zip)

**Reply by js-jixu** [reply]
Forgive me that I didn't fully understand the meaning of your words before. Now I know that Σr=Σag+Σsg->g in Moltres's setting and I can get the same results as you. The difference between Moltres solution and the reference solution is 449 pcm . Although this difference is not large, other codes using the block method calculate results that are only a few pcm from the reference solution, the comparision of results calculated by different codes are attached. What do you think of this result？ My tutor said that maybe the size of mesh is too large for the calculation of neutron diffusion equation. He wrote a code which is capable of calculating neutron diffusion equation some years ago. He said that when he used a denser grid the results were quite different from the reference solution, but when he used a coarser grid he got very similar results to the reference solution.

![comparision](https://user-images.githubusercontent.com/99813361/167262460-c5333ca2-67e7-406c-9643-d34a6296992e.png)

**Reply by js-jixu** [reply]
Hi, Sun Myung. 

I also wonder that whether you done any benchmarks to demonstrate the capability of Moltres to do neutron dynamics calculation? 

At the suggestion of my tutor, I'm doing three qualitative analyses to demonstrate the capability of Moltres to do neutron dynamics. He said that if I can get reasonable results, he thinks Moltres is capable of doing neutron dynamics analysis. But if I want to prove that, there is no other way than to use Moltres to do a benchmark. The three qualitative analyses are time step sensitivity, point reactor dynamics analysis and critical single rod analysis. I have finished the first two, and the third one is in the process of being done, and it is expected that the single rod of the lead-cooled reactor will be adjusted to the critical state tomorrow. The results of the first two is very good, I'll let you know the third one when I finish the third one. 

Besides that, I also want to do a neutron dynamics benchmark to demonstrate Moltres, do you have any advice?

With best wishes,

Xu.

**Reply by smpark7** [reply]
Hey Xu. I think you'll get very close to the reference result if you include the axial buckling leakage as mentioned by the IAEA benchmark paper (B^2=0.8E-4). You can do this by including an additional DB^2 term in the removal cross section as shown in this [paper](https://www.researchgate.net/profile/Tagor-Sembiring/publication/266279695_Application_of_Neutron_Diffusion_Theory_for_High_Uranium_Density_Fueled_Reactor_Core_Calculation/links/54fd8b9a0cf270426d12987b/Application-of-Neutron-Diffusion-Theory-for-High-Uranium-Density-Fueled-Reactor-Core-Calculation.pdf).

For the reactor dynamics analysis, it depends on what you want to simulate. Our [verification](https://www.sciencedirect.com/science/article/pii/S0306454922001463) with the CNRS benchmark has a time-dependent case studying the response in power output from perturbations in the heat removal rate and reactivity feedback. The input files are in `moltres/problems/2021-cnrs-benchmark`

**Reply by js-jixu** [reply]
Thanks a lot, Sun Myung. I'll read the paper carefully. And congratulations on your published paper!!!
